Add an option `WOODPECKER_ROOT_URL` to let the user set a path prefix

restray opened the pull request:
closes #1636 
closes #1429

Hey!

I wanted to be able to set a path prefix (behind a reverse proxy) and it wouldn't work.

Hope that you'll adopt it :)

6543 commented on pipeline/rpc/proto/woodpecker_grpc.pb.go:
```diff
@@ -8,6 +8,7 @@ package proto
 
 import (
 	context "context"
+
```
Please dont touch codegenerated files

6543 commented on pipeline/rpc/proto/woodpecker.pb.go:
```diff
@@ -7,10 +7,11 @@
 package proto
 
 import (
```
Please dont touch codegenerated files

6543 commented on server/forge/mocks/forge.go:
```diff
@@ -34,7 +34,7 @@ func (_m *Forge) Activate(ctx context.Context, u *model.User, r *model.Repo, lin
 }
 
 // Auth provides a mock function with given fields: ctx, token, secret
-func (_m *Forge) Auth(ctx context.Context, token string, secret string) (string, error) {
```
Please dont touch codegenerated files

6543 commented on cmd/server/flags.go:
```diff
@@ -45,6 +45,11 @@ var flags = []cli.Flag{
 		Usage:   "server fully qualified url (<scheme>://<host>)",
 	},
+	&cli.StringFlag{
+		EnvVars: []string{"WOODPECKER_ROOT_URL"},
```
docs are mssing

restray replied:
It should be good now!

6543 commented:
@restray did fix the merge conflicts and adressed the rename for you ... if you need help just tell

restray commented:
> @restray did fix the merge conflicts and adressed the rename for you ... if you need help just tell

Thanks you very much! I come back later :D

restray commented on web/build.html:
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
```
The index.html is used when you develop `pnpm run start`!

This file is used to be build.

6543 commented on server/web/config.go:
```diff
@@ -58,6 +59,7 @@ func Config(c *gin.Context) {
 	c.Header("Content-Type", "text/javascript; charset=utf-8")
 	tmpl := template.Must(template.New("").Funcs(funcMap).Parse(configTemplate))
 
+	c.Status(http.StatusOK)
```
I would add the status after the error handeling (raw http std would have an issue - nut sure if gin do ...)

restray commented:
Hey! I've reviewed, modified and rebased from master (with a little mistakes)! Everything should be alright :) I'm waiting for your reviews! Thanks for the time you're spending on this PR!

6543 commented:
I still think that the index.html is not the best sollution ... I'll have a look if i can do that in some go-template fashion or so
